### PR TITLE
Record refunds separate from order entries

### LINKED ISSUE #2001: Refunds should be recorded on the date refunded
**Describe the bug**
Currently refunds are tracked on the original order entry in the `wc_order_stats` table.  If a refund is provided a month later, the net total is recorded on the original order data and not on the date refunded.

Example: $100 order on 6/1 and refund is provided on 7/1.
**Current**: $50 net amount is shown on 6/1.
**Expected**: $100 is shown on 6/1 and -$50 is shown on 7/1.

This PR will involve:
* Recording refunds as their own row in the `wc_order_stats` table.
* Adding a `parent_id` field to the `wc_order_stats` table to track refunds/parent order relationships.
* Possibly removing the `refund_amount` field as this might be redundant once we store refunds.

More discussion is available at p6riRB-46r-p2

### PR DESCRIPTION
**Pull request by joshuatf**
Fixes #2001 

Records refunds in their own rows so that stats are reflected on the date the refund is given and not on the original order date.

* Reflects refunded amounts in orders, products/categories, taxes, and revenue.
* Removes unused refund fields no longer needed since refunds are now recorded.
* Updates tests to include refunded amounts during a period.
* Removes refunded orders from excluded statuses since they are needed to balance out the actual refund amounts.

### Thoughts

1. There’s some code duplication between the `WC_Admin_Order` and `WC_Admin_Order_Refund` classes.  We can’t use traits since we need to support PHP 5.2.  Any other thoughts on how to clean this up?
2. Should the refund count go towards the `avg_order_value`?  `( net_total / num_orders )` - Should refunds be included in the number of orders here?  This PR assumes they go towards the net total, but don't influence the order count `( net total of all orders and refunds ) / ( number of actual orders excluding refunds )`.

### Screenshots
Screen Shot 2019-04-22 at 4 29 13 PM
Screen Shot 2019-04-22 at 5 03 31 PM

### Detailed test instructions:

1.  Create an order with multiple products, taxes, and at least one coupon.
2.  Make a couple partial refunds to that order.
3.  Make sure the refunds show up correctly in each report with negative values when the refund is added.
4.  Make sure all tests pass.

### CODE REVIEW

**Review by psealock — COMMENTED** [review]
This is great @joshuatf. I only had a chance to briefly test this so far.

Looking at the orders report, refunds are getting added with "New" customer status

Screen Shot 2019-04-25 at 11 54 05 AM

I would expect "Returning", nothing at all, or maybe "N/A".

**Comment by timmyc** [comment]
> We can’t use traits since we need to support PHP 5.2.

Since WP Core has moved the [minimum PHP version to 5.5](https://core.trac.wordpress.org/changeset/44950) it seems like we could explore cleaning things up with traits.

> Should the refund count go towards the avg_order_value? ( net_total / num_orders )

I seem to think this is the correct path to take based on my conversations with @LevinMedia on the topic.

> Should refunds be included in the number of orders here? This PR assumes they go towards the net total, but don't influence the order count ( net total of all orders and refunds ) / ( number of actual orders excluding refunds ).

This seems reasonable to me too.

**Review by timmyc — COMMENTED** [review]
Impressive work here @joshuatf - I have to admit I'm not an expert in how Woo handles returns so I'm going to cc some folks on Proton to give this one some 👀 too

**Comment by timmyc** [comment]
@peterfabian / @claudiosanches any chance you could give this a review?

File: includes/class-wc-admin-order-refund.php [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	 * @param WC_Order_Item $item Line item from order.
+	 *
+	 * @return float|int
+	 */
+	public function get_item_shipping_amount( $item ) {
+		// Shipping amount loosely based on woocommerce code in includes/admin/meta-boxes/views/html-order-item(s).php
+		// distributed simply based on number of line items.
+		$product_qty = $item->get_quantity( 'edit' );
+		$order_items = $this->get_item_count();
+		if ( 0 === $order_items ) {
```
[the remark this answers is not in the document]
Comment (joshuatf) [reply includes/class-wc-admin-order-refund.php]: @timmyc Is it ever possible for a product quantity to be 0?  I thought all line items required a minimum of 1 quantity, but I may be mistaken.

**Comment by joshuatf** [comment]
Thanks for the review @psealock and @timmyc!

>Since WP Core has moved the minimum PHP version to 5.5 it seems like we could explore cleaning things up with traits.

Sweet 🎉 - Updated this to a minimum of php 5.4 used in [WPCS](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/tree/develop), but if you think this should be 5.5 let me know and I'll update accordingly.

>Should the refund count go towards the avg_order_value? ( net_total / num_orders )

@LevinMedia Can you confirm that this looks okay when you have time?

>Looking at the orders report, refunds are getting added with "New" customer status.  I would expect "Returning", nothing at all, or maybe "N/A".

I think you're right.  We're not counting these towards returning in the database so nothing or `N/A` is probably appropriate here.  I'll add that to [this PR](https://github.com/woocommerce/woocommerce-admin/pull/2102).

**Comment by psealock** [comment]
>Updated this to a minimum of php 5.4 used in WPCS, but if you think this should be 5.5 let me know and I'll update accordingly.

This means we'll require WordPress 5.2, so we may as well change to PHP v5.5.

>Should the refund count go towards the avg_order_value? ( net_total / num_orders )

I would think so. 

This one is working well for me. I'm inclined to approve, but a 👍 from @claudiosanches or @peterfabian would be helpful.

**Comment by claudiosanches** [comment]
This code looks good, but I'll leave for Peter the decision about PHP/WP requirements.

**Review by psealock — APPROVED** [review]
Besides the PHP requirement issue this is looking good to me and working. With @claudiosanches signing off, we can tackle the PHP issue separately.

**Comment by joshuatf** [comment]
Thanks for the review @psealock 😄 

Will open an issue regarding any updates for PHP/WP requirements so we can get this merged in.

**Comment by claudiosanches** [comment]
I like the idea of writing PHP 5.6 code for sure!
Just need to check with Proton.

**Pull request merged**